### PR TITLE
Get BDF.jl to pass

### PR DESCRIPTION
**Pull request by calebwin**
This PR added a bunch of things to finally get BDF.jl tests to pass:

- Handle scheduling failure by destroying the job
- Handle empty divisions - changes to `ReadGroup`, `SplitGroup`, and `Shuffle`
- Added to the docs to explain that aggregations should probably have an `init` parameter set in case some partition is empty

### CODE REVIEW

File: Banyan/src/jobs.jl [hunk 1]
```diff
@@ -189,6 +189,11 @@ function destroy_job(job_id::JobId = get_job_id(); failed = nothing, force = fal
     )
 
     # Remove from global state
+    # TODO: Maybe `job_id`must always has be in `jobs`. In that case, we should
+    # not be needing this `if` statement here.
+    if failed == true && haskey(jobs, job_id)
```
Comment (cailinw) [inline Banyan/src/jobs.jl]:
```suggestion
    if failed && haskey(jobs, job_id)
```

File: BanyanDataFrames/test/groupby_filter_indexing.jl [hunk 2]
```diff
@@ -218,176 +218,176 @@ end
 
         bucket = get_cluster_s3_bucket_name()
 
-        for i = 1:2
+        for i = 1:1#2
```
Comment (cailinw) [inline BanyanDataFrames/test/groupby_filter_indexing.jl]: Do you want to commit this change? By reducing the loop bounds to 1:1, this is not testing `write_csv`, etc.

File: BanyanDataFrames/test/groupby_filter_indexing.jl [hunk 3]
```diff
@@ -218,176 +218,176 @@ end
...
-                    4.26,
-                    4.26,
-                    5.55,
-                    5.55,
-                    5.55,
-                    5.55,
-                    5.55,
-                    5.55,
-                ]
+                # gdf_select_save_path = "s3://$(bucket)/test-tmp_gdf_select_$(split(path, "/")[end])"
```
Comment (cailinw) [inline BanyanDataFrames/test/groupby_filter_indexing.jl]: Do you want to keep these tests commented out? is it indicative of certain tests not passing as yet?

File: BanyanDataFrames/test/runtests.jl [hunk 4]
```diff
@@ -60,7 +60,8 @@ function use_job_for_testing(
                     "banyan-julia/BanyanArrays",
                     "banyan-julia/BanyanDataFrames"
                 ],
-                force_pull = true,
+                force_pull = get(ENV, "BANYAN_FORCE_RECLONE", "0") == "0",
+                force_reclone = get(ENV, "BANYAN_FORCE_RECLONE", "0") == "1",
```
Comment (cailinw) [inline BanyanDataFrames/test/runtests.jl:64]: Nice, thanks. Was it intended to check `"BANYAN_FORCE_RECLONE"` for both these parameters? That is, it was always either force_pull or force_reclone?

File: Banyan/src/jobs.jl [hunk 5]
```diff
@@ -189,6 +189,11 @@ function destroy_job(job_id::JobId = get_job_id(); failed = nothing, force = fal
...
+        jobs[job_id].current_status = "failed"
```
Comment (cailinw) [inline Banyan/src/jobs.jl]: Is there anywhere in the code that is looking at the global `jobs` and checking the status? I recall we were in the past, but I thought that this wasn't needed now.

**Review by cailinw — CHANGES_REQUESTED** [review]
Looks good, thank you. Do you want to retain all the prints in this PR?

File: BanyanDataFrames/test/runtests.jl (hunk 4 above)
Comment (calebwin) [reply BanyanDataFrames/test/runtests.jl:64]: It only reclones if `BANYAN_FORCE_RECLONE=1`.

File: BanyanDataFrames/test/groupby_filter_indexing.jl (hunk 2 above)
Comment (calebwin) [reply BanyanDataFrames/test/groupby_filter_indexing.jl]: See below comment.

File: Banyan/src/jobs.jl (hunk 5 above)
Comment (calebwin) [reply Banyan/src/jobs.jl]: No it's not used anywhere. I deleted it.

File: BanyanDataFrames/test/groupby_filter_indexing.jl (hunk 3 above)
Comment (calebwin) [reply BanyanDataFrames/test/groupby_filter_indexing.jl]: They are passing, forgot to commit.

**Review by cailinw — APPROVED** [review]

**Pull request merged**